### PR TITLE
Refactor deadman set failmode to be cross platform

### PR DESCRIPTION
**Pull request by mattmacy**
Signed-off-by: Matt Macy <mmacy@FreeBSD.org>

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/os/linux/zfs/spa_misc_os.c [hunk 1]
```diff
@@ -42,6 +42,17 @@
...
 
+
+int
+param_set_deadman_failmode(const char *val, zfs_kernel_param_t *kp)
+{
+	int error;
+
+	error = param_set_deadman_failmode_common(val);
+
+	return (error || param_set_charp(val, kp));
```
Comment (behlendorf) [inline module/os/linux/zfs/spa_misc_os.c]: This isn't quite right, we need the return value from `param_set_charp()` returned.  Let's keep it simple:

```
        int error = -param_set_deadman_failmode_common(val);
        if (error == 0)
                error = param_set_charp(val, kp);

        return (error);
```

File: module/zfs/spa_misc.c [hunk 2]
```diff
@@ -2731,7 +2731,7 @@ param_set_deadman_failmode(const char *val, zfs_kernel_param_t *kp)
 		mutex_exit(&spa_namespace_lock);
 	}
 
-	return (param_set_charp(val, kp));
```
Comment (behlendorf) [inline module/zfs/spa_misc.c]: We'll want the common function to term positive errno's on lines 2718 and 2725.

File: module/os/linux/zfs/spa_misc_os.c [hunk 3]
```diff
@@ -42,6 +42,19 @@
 #include <sys/kstat.h>
 #include "zfs_prop.h"
 
+
+int
+param_set_deadman_failmode(const char *val, zfs_kernel_param_t *kp)
```
Comment (behlendorf) [inline module/os/linux/zfs/spa_misc_os.c]: The `module_param_call` function should be switched to use the `ZFS_MODULE_PARAM_CALL` wrapper.

```diff
diff --git a/module/zfs/spa_misc.c b/module/zfs/spa_misc.c
index 3956fc4..6af6c95 100644
--- a/module/zfs/spa_misc.c
+++ b/module/zfs/spa_misc.c
@@ -2847,11 +2847,9 @@ ZFS_MODULE_PARAM(zfs, zfs_, ddt_data_is_special, INT, ZMO
 ZFS_MODULE_PARAM(zfs, zfs_, user_indirect_is_special, INT, ZMOD_RW,
        "Place user data indirect blocks into the special class");
 
-#ifdef _KERNEL
-module_param_call(zfs_deadman_failmode, param_set_deadman_failmode,
-       param_get_charp, &zfs_deadman_failmode, 0644);
-MODULE_PARM_DESC(zfs_deadman_failmode, "Failmode for deadman timer");
-#endif
+ZFS_MODULE_PARAM_CALL(zfs_deadman, zfs_deadman_, failmode,
+       param_set_deadman_failmode, param_get_charp, ZMOD_RW,
+       "Failmode for deadman timer");
 
 /* BEGIN CSTYLED */
 ZFS_MODULE_PARAM_CALL(zfs_deadman, zfs_deadman_, synctime_ms,
```

File: include/sys/spa_impl.h [hunk 4]
```diff
@@ -433,6 +433,7 @@ struct spa {
 };
 
 extern char *spa_config_path;
+extern char *zfs_deadman_failmode;
```
Comment (behlendorf) [inline include/sys/spa_impl.h]: This can be removed and `zfs_deadman_failmode` can remain private.  It's only accessed from within `spa_misc.c`.

Comment (mattmacy) [reply include/sys/spa_impl.h]:
```
static int 
zfs_deadman_failmode_sysctl(SYSCTL_HANDLER_ARGS)
{
	char buf[16];
	int rc;

	if (req->newptr == NULL)
		strlcpy(buf, zfs_deadman_failmode, sizeof (buf));

	rc = sysctl_handle_string(oidp, buf, sizeof (buf), req);
	if (rc || req->newptr == NULL)
		return (rc);
	if (strcmp(buf, zfs_deadman_failmode) == 0)
		return (0);
	if (!strcmp(buf,  "wait"))
		zfs_deadman_failmode = "wait";
	if (!strcmp(buf,  "continue"))
		zfs_deadman_failmode = "continue";
	if (!strcmp(buf,  "panic"))
		zfs_deadman_failmode = "panic";

	return (-param_set_deadman_failmode_common(buf));
}
```

Comment (behlendorf) [reply include/sys/spa_impl.h]: Ahh, ok.

**Pull request merged**